scheduler/pkg/klcpermit/workflow_manager.go

```diff
@@ -54,73 +56,95 @@ type Manager interface {
-	}
+	sMgr.startWatching(ctx, informer.Informer(), pod, handler)
+}
+
+func (sMgr *WorkloadManager) startWatching(ctx context.Context, s cache.SharedIndexInformer, pod *corev1.Pod, handler framework.WaitingPod) {
+	stopCh := make(chan struct{})
+	workloadInstanceName := getCRDName(pod)
+
+	checkWorkloadInstance := func(obj interface{}) {
+		unstructuredWI := obj.(*unstructured.Unstructured)
 
-	_, span := sMgr.getSpan(ctx, crd, pod)
-
-	//check CRD status
-	phase, found, err := unstructured.NestedString(crd.UnstructuredContent(), "status", "preDeploymentEvaluationStatus")
-	klog.Infof("[Keptn Permit Plugin] workloadInstance crd %s, found %s with phase %s ", crd, found, phase)
-	if err == nil && found {
-		span.AddEvent("StatusEvaluation", trace.WithAttributes(tracing.Status.String(phase)))
-		switch KeptnState(phase) {
-		case StateFailed:
-			span.SetStatus(codes.Error, "Failed")
-			span.End()
-			unbindSpan(pod)
-			return Failure
-		case StateSucceeded:
-			span.End()
-			unbindSpan(pod)
-			return Success
-		case StatePending:
-			return Wait
-		case StateRunning:
-			return Wait
-		case StateUnknown:
-			return Wait
+		if unstructuredWI.GetName() != workloadInstanceName {
+			return
 		}
-	}
-	return WorkloadInstanceStatusNotSpecified
-}
 
-// GetCRD returns unstructured to avoid tight coupling with the CRD resource
-func (sMgr *WorkloadManager) GetCRD(ctx context.Context, namespace string, name string) (*unstructured.Unstructured, error) {
-	// GET /apis/lifecycle.keptn.sh/v1/namespaces/{namespace}/workloadinstance/name
-	return sMgr.dynamicClient.Resource(workloadInstanceResource).Namespace(namespace).Get(ctx, name, metav1.GetOptions{})
+		_, span := sMgr.getSpan(ctx, unstructuredWI, pod)
+
+		phase, found, err := unstructured.NestedString(unstructuredWI.UnstructuredContent(), "status", "preDeploymentEvaluationStatus")
+		if err != nil {
+			klog.Errorf("[Keptn Permit Plugin] cannot fetch workloadInstance crd %s preDeploymentStatus: %s", unstructuredWI, err.Error())
+			return
+		}
+		klog.Infof("[Keptn Permit Plugin] workloadInstance crd %s, found %s with phase %s ", unstructuredWI, found, phase)
+		if found {
+			span.AddEvent("StatusEvaluation", trace.WithAttributes(tracing.Status.String(phase)))
+			switch KeptnState(phase) {
+			case StateFailed, StateCancelled:
+				handler.Reject(PluginName, "Pre Deployment Check failed")
+				span.End()
+				sMgr.unbindSpan(pod)
+				stopCh <- struct{}{}
+			case StateSucceeded:
+				handler.Allow(PluginName)
+				span.End()
+				sMgr.unbindSpan(pod)
+				stopCh <- struct{}{}
+			case StatePending:
+			case StateRunning:
+			case StateUnknown:
+			}
+		}
+	}
+	handlers := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			checkWorkloadInstance(obj)
+		},
+		UpdateFunc: func(oldObj, obj interface{}) {
+			checkWorkloadInstance(obj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			klog.Info("received delete event!")
+		},
+	}
+	s.AddEventHandler(handlers)
+	s.Run(stopCh)
```

Comment:
I do not really like directly using the run method here, since we are using the dynamic factory we could use 
	factory.Start(stopCh)
	factory.WaitForCacheSync(stopCh)
Also this requires changing our code to remove the go routine since this is already created by the start function

Reply:
usually we should wait for cache sync to avoid reasoning on a cache which is not uptodate at pod shutdown/restart

Reply:
we need to keep it in go-routine, otherwise it does not work and schedule pods

Reply:
Start has already a go routine where the concurrency logic is handledby the informer library tho